### PR TITLE
store: implement serveArchiveFile API.

### PR DESCRIPTION
**Pull request by frankban**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/87/

File: internal/v4/archive.go [hunk 1]
```diff
@@ -128,6 +125,48 @@ func verifyConstraints(s string) error {
...
+		}
+		defer content.Close()
+		// Send the response to the client.
+		ctype := mime.TypeByExtension(filepath.Ext(filePath))
+		if ctype != "" {
+			w.Header().Set("Content-Type", ctype)
+		}
+		w.Header().Set("Content-Length", strconv.FormatInt(fileInfo.Size(), 10))
+		w.WriteHeader(http.StatusOK)
+		io.Copy(w, content)
```
Comment (jrwren) [inline internal/v4/archive.go:164]: Should return value be checked here comparing to content length? Should err be logged if there is one?

**Comment by jrwren** [comment]
:+1:

File: internal/v4/archive.go [hunk 2]
```diff
@@ -152,8 +191,20 @@ func (h *handler) nextRevisionForId(id *charm.Reference) (int, error) {
 	return 0, nil
 }
 
-// GET id/archive/…
-// http://tinyurl.com/lampm24
-func (h *handler) serveArchiveFile(charmId *charm.Reference, w http.ResponseWriter, req *http.Request) error {
-	return errNotImplemented
+func (h *handler) openBlobById(id *charm.Reference) (blobstore.ReadSeekCloser, int64, error) {
```
Comment (rogpeppe) [inline internal/v4/archive.go]: just openBlob, perhaps? I'm not sure how else you might open one except by id.

File: internal/v4/archive_test.go [hunk 3]
```diff
@@ -368,6 +370,98 @@ func (s *ArchiveSuite) assertUpload(c *gc.C, url *charm.Reference, fileName stri
...
+	// Retrieve the expected bytes.
+	var expectBytes []byte
+	for _, file := range zipFile.File {
+		if file.Name == filePath {
+			r, err := file.Open()
+			c.Assert(err, gc.IsNil)
+			defer r.Close()
+			expectBytes, err = ioutil.ReadAll(r)
+			c.Assert(err, gc.IsNil)
+		}
```
Comment (rogpeppe) [inline internal/v4/archive_test.go:449]: break?

File: internal/v4/archive.go [hunk 4]
```diff
@@ -128,6 +125,48 @@ func verifyConstraints(s string) error {
...
+		return err
+	}
+	defer r.Close()
+	zipReader, err := zip.NewReader(&readerAtSeeker{r}, size)
+	if err != nil {
+		return errgo.Notef(err, "cannot read archive data for %s", id)
+	}
+
+	// Retrieve the requested file from the zip archive.
+	filePath := strings.TrimPrefix(req.URL.String(), "/")
```
Comment (rogpeppe) [inline internal/v4/archive.go]: This seems a bit odd. Doesn't req.URL.String() start with http:// ?
Shouldn't it be req.URL.Path ? I'm not sure quite how it's working right now.
Hmm, looks like the URL doesn't have a method. Still better
to  use req.URL.Path though, I think.

Also, perhaps path.Clean the URL path?

File: internal/v4/archive.go [hunk 5]
```diff
@@ -128,6 +125,48 @@ func verifyConstraints(s string) error {
...
+	for _, file := range zipReader.File {
+		if path.Clean(file.Name) != filePath {
+			continue
+		}
```
Comment (rogpeppe) [inline internal/v4/archive.go:146]: Do we want to TrimPrefix("./") perhaps? I remember seeing some logic
that did that once, and it makes some sense.

File: internal/v4/archive.go (hunk 1 above)
Comment (rogpeppe) [inline internal/v4/archive.go:164]: I think this is OK as is. The content can be truncated for a few reasons (e.g. client going down),
and I'm not sure that logging that message is likely to be that useful.

**Comment by rogpeppe** [comment]
LGTM with a few minor suggestion. Nice, thanks.

File: internal/v4/archive.go (hunk 2 above)
Comment (frankban) [inline internal/v4/archive.go]: By hash, by entity? [shrug]
I don't have a strong opinion, so I guess I'll change this as you suggested.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/93/

File: internal/v4/archive_test.go (hunk 3 above)
Comment (frankban) [inline internal/v4/archive_test.go:449]: Oh, good catch, fixed.

File: internal/v4/archive.go (hunk 4 above)
Comment (frankban) [inline internal/v4/archive.go]: It doesn't, but I agree with you it's better to use URL.Path here.
Done.

File: internal/v4/archive.go (hunk 5 above)
Comment (frankban) [inline internal/v4/archive.go:146]: path.Clean does that automatically.

**Comment by frankban** [comment]
Thanks for the reviews!
:shipit:

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/94/

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**